x/ecocredit/server/core/create_batch.go

```diff
@@ -0,0 +1,146 @@
+package core
+
+import (
+	"context"
+	"github.com/cosmos/cosmos-sdk/orm/types/ormerrors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ecocreditv1beta1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1beta1"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	"github.com/regen-network/regen-ledger/x/ecocredit/v1beta1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// CreateBatch creates a new batch of credits.
+// Credits in the batch must not have more decimal places than the credit type's specified precision.
+func (k Keeper) CreateBatch(ctx context.Context, req *v1beta1.MsgCreateBatch) (*v1beta1.MsgCreateBatchResponse, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	projectID := req.ProjectId
+
+	projectInfo, err := k.stateStore.ProjectInfoStore().GetByName(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	classInfo, err := k.stateStore.ClassInfoStore().Get(ctx, projectInfo.ClassId)
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.assertClassIssuer(ctx, projectInfo.ClassId, req.Issuer)
+	if err != nil {
+		return nil, err
+	}
+
+	batchSeqNo, err := k.getBatchSeqNo(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	batchDenom, err := ecocredit.FormatDenom(classInfo.Name, batchSeqNo, req.StartDate, req.EndDate)
+	if err != nil {
+		return nil, err
+	}
+
+	rowID, err := k.stateStore.BatchInfoStore().InsertReturningID(ctx, &ecocreditv1beta1.BatchInfo{
+		ProjectId:  projectInfo.Id,
+		BatchDenom: batchDenom,
+		Metadata:   req.Metadata,
+		StartDate:  timestamppb.New(req.StartDate.UTC()),
+		EndDate:    timestamppb.New(req.EndDate.UTC()),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var p ecocredit.Params
+	k.params.GetParamSet(sdkCtx, &p)
+	creditType, err := k.getCreditType(classInfo.CreditType, p.CreditTypes)
+	if err != nil {
+		return nil, err
+	}
+	maxDecimalPlaces := creditType.Precision
+
+	tradableSupply, retiredSupply := math.NewDecFromInt64(0), math.NewDecFromInt64(0)
+	for _, issuance := range req.Issuance {
+		decs, err := getNonNegativeFixedDecs(maxDecimalPlaces, issuance.TradableAmount, issuance.RetiredAmount)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
at the core it would do `return Dec{}, fmt.Errorf("%s exceeds maximum decimal places: %d", s, max)`, where `s` is the passed decimal.  should we get more specific than that? i'm not sure if theres a simple way to tell the user specifically whether it was their tradable or retired amount

Reply:
What about wrapping this error with `invalid tradable or retired amount`? This gets more complicated in other uses but the user provided inputs should probably be included in the error message so the user knows what needs to be fixed.